### PR TITLE
[RTM] Multi Threading for time communication 

### PR DESCRIPTION
**Pull request by benjicanal**
Answering issue number 3 #3

### CODE REVIEW

File: pyScienceMode2/rehastim_interface.py [hunk 1]
```diff
@@ -332,36 +323,20 @@ def stop_stimulation(self):
...
-        ----------
-        stimulation_duration: float
-            Time of the stimulation after the update.
-        upd_list_channels: list[Channel]
-            List of the channels that will be updated
         """
         self.amplitude = [0] * len(self.list_channels)
         self._send_packet("StartChannelListMode")
+        self._get_last_ack()
+        self.stimulation_started = True
```
Comment (aceglia) [inline pyScienceMode2/rehastim_interface.py:330]: stimulation_started = False ?

File: pyScienceMode2/rehastim_interface.py [hunk 2]
```diff
@@ -332,36 +323,20 @@ def stop_stimulation(self):
...
         Warning: only the channel that has been initiated can be updated.
-
-        Parameters
-        ----------
-        stimulation_duration: float
-            Time of the stimulation after the update.
-        upd_list_channels: list[Channel]
-            List of the channels that will be updated
         """
         self.amplitude = [0] * len(self.list_channels)
```
Comment (aceglia) [inline pyScienceMode2/rehastim_interface.py:327]: Store amplitude in a temporary variable to reset self.amplitude to initial values: 
tmp_amp = self.amplitude
self.amplitude = 0
self._sendpacket
self._get_last_ack
self.amplitude = tmp_amp

File: pyScienceMode2/sciencemode.py [hunk 3]
```diff
@@ -7,7 +7,7 @@
 import time
 import threading
 from pyScienceMode2.utils import *
-from .acks import motomed_error_ack
+from .acks import *
```
Comment (aceglia) [inline pyScienceMode2/sciencemode.py:10]: Try to not import all but import the specific class instead.

File: pyScienceMode2/sciencemode.py [hunk 4]
```diff
@@ -75,30 +75,36 @@ def __init__(self, port: str, show_log: bool = False, with_motomed: bool = False
...
         self.is_phase_result = threading.Event()
-        self.command_sent = threading.Event()
         self.event_ack = threading.Event()
-        self.__motomed_thread_started = False
+        self.last_phase_result = None
+        self._motomed_command_done = True
+        self.is_motomed_connected = with_motomed
+        self.__comparison_thread_started = False
         self.__watchdog_thread_started = False
+        self.info_send = []  # Command sent to the rehastim
```
Comment (aceglia) [inline pyScienceMode2/sciencemode.py:96]: if it is a command then call it command not info.

File: pyScienceMode2/sciencemode.py [hunk 5]
```diff
@@ -118,66 +124,139 @@ def _get_last_ack(self, init: bool = False) -> bytes:
...
-    def _start_motomed_thread(self):
+        Returns
+        -------
+        self.info_received : list
+            Commands received from the rehastim
         """
-        Start the thread which catch motomed data.
+        return self.info_received
+
+    def return_list_send(self) -> list:
```
Comment (aceglia) [inline pyScienceMode2/sciencemode.py:162]: Is it a private function ?

Comment (aceglia) [reply pyScienceMode2/sciencemode.py:162]: More explicit on the function name

File: pyScienceMode2/sciencemode.py [hunk 6]
```diff
@@ -118,66 +124,139 @@ def _get_last_ack(self, init: bool = False) -> bytes:
...
+                if self.show_log and packet[-1][6] in [t.value for t in self.Type]:
+                    print(f"Ack received by rehastim: {self.Type(packet[-1][6]).name}")
+                    self.info_received.append(packet[-1])
+                    self.return_list_ack()
+        if self.error_occured:
+            raise RuntimeError("Stimulation error")
+
+        return packet[-1]
+
+    def return_list_ack(self) -> list:
```
Comment (aceglia) [inline pyScienceMode2/sciencemode.py:151]: Is it a private function ?
More explicit on the function name

File: pyScienceMode2/sciencemode.py [hunk 7]
```diff
@@ -118,66 +124,139 @@ def _get_last_ack(self, init: bool = False) -> bytes:
...
+        self.info_send : list
+            Commands sent to the rehastim
         """
-        self.__motomed_thread_started = True
-        self.__thread_motomed = threading.Thread(target=self._catch_motomed_data)
-        self.__thread_motomed.start()
+        return self.info_send
 
-    def _catch_motomed_data(self):
+    def _start_thread_catch_data(self):
```
Comment (aceglia) [inline pyScienceMode2/sciencemode.py:173]: thread_catch_ack can be more relevant.

File: pyScienceMode2/sciencemode.py [hunk 8]
```diff
@@ -118,66 +124,139 @@ def _get_last_ack(self, init: bool = False) -> bytes:
...
         """
-        Catch the motomed data.
+        Start the thread which catches rehastim data and motomed data if motomed flag is true.
         """
+        self.__comparison_thread_started = True
+        self.__thread_catch_data = threading.Thread(target=self._thread_catch_data)
+        self.__thread_catch_data.start()
+
+    def _thread_catch_data(self):
```
Comment (aceglia) [inline pyScienceMode2/sciencemode.py:181]: idem catch acks

**Review by aceglia — CHANGES_REQUESTED** [review]

File: pyScienceMode2/rehastim_interface.py (hunk 1 above)
Comment (benjicanal) [reply pyScienceMode2/rehastim_interface.py]: removed because useless

File: pyScienceMode2/sciencemode.py (hunk 3 above)
Comment (benjicanal) [reply pyScienceMode2/sciencemode.py]: Done.

File: pyScienceMode2/sciencemode.py (hunk 4 above)
Comment (benjicanal) [reply pyScienceMode2/sciencemode.py]: Renamed

File: pyScienceMode2/sciencemode.py (hunk 6 above)
Comment (benjicanal) [reply pyScienceMode2/sciencemode.py]: Yes, done.

File: pyScienceMode2/sciencemode.py (hunk 5 above)
Comment (benjicanal) [reply pyScienceMode2/sciencemode.py]: Done.

File: pyScienceMode2/sciencemode.py (hunk 7 above)
Comment (benjicanal) [reply pyScienceMode2/sciencemode.py]: Done.

File: pyScienceMode2/sciencemode.py (hunk 8 above)
Comment (benjicanal) [reply pyScienceMode2/sciencemode.py]: Done.

File: pyScienceMode2/rehastim_interface.py (hunk 2 above)
Comment (benjicanal) [reply pyScienceMode2/rehastim_interface.py:328]: Done.

**Review by benjicanal — COMMENTED** [review]
Correct the comment review

**Comment by aceglia** [comment]
Is the code blacked ? If not please do it and I can accept the PR

**Comment by benjicanal** [comment]
> Is the code blacked ? If not please do it and I can accept the PR

Yes the code is blacked.

**Pull request merged**